### PR TITLE
Add valid for days

### PR DESCRIPTION
**Pull request by FadhlanR**

### CODE REVIEW

**Review by jurgenwerk — CHANGES_REQUESTED** [review]
Can you add a PR description with some more info about this change? Since the recurring day can be different from the amount of days in the month it seems that the check has several edge cases we need to take care of. Would be great if you could add comments describing edge cases that were considered and tested. 

I think it would also be beneficial to have a ticket for this so we can track progress on when this gets staged or deployed.

File: test/Config.spec.ts [hunk 1]
```diff
@@ -8,6 +8,7 @@ describe("Config", async () => {
   const [user1, user2, user3] = waffle.provider.getWallets();
   const crankAddress = user2.address;
   const feeReceiver = user3.address;
+  const validForDays = 259200; //3 days
```
Comment (alex-cardstack) [inline test/Config.spec.ts:11]: Ah, it's not stored as days. I think this is a bit of a footgun - I think maybe we should rename it to `validForSeconds` in the contract code

Comment (jurgenwerk) [reply test/Config.spec.ts:9]: I'd keep it in days, i.e. `validForDays = 3`, this is how we do it in the hub also

File: contracts/ScheduledPaymentModule.sol [hunk 2]
```diff
@@ -281,6 +279,29 @@ contract ScheduledPaymentModule is Module {
...
+                block.timestamp / BokkyPooBahsDateTimeLibrary.SECONDS_PER_DAY
+            );
+        uint256 recursDate = BokkyPooBahsDateTimeLibrary.timestampFromDate(
+            year,
+            month,
+            recursDayOfMonth
+        );
+
+        return
+            block.timestamp.sub(_lastPaidAt) >= 28 days && //recursDayOfMonth value range 1-28
```
[the remark this answers is not in the document]
Comment (jurgenwerk) [reply contracts/ScheduledPaymentModule.sol]: See this comment: https://github.com/cardstack/cardstack/pull/3220#discussion_r970995395 (this is how this is now implemented in the hub)

File: contracts/Config.sol [hunk 3]
```diff
@@ -33,4 +36,8 @@ contract Config is IConfig, OwnableUpgradeable {
     function getFeeReceiver() external view returns (address) {
         return feeReceiver;
     }
+
+    function getValidForDays() external view returns (uint256) {
```
[the remark this answers is not in the document]
Comment (FadhlanR) [reply contracts/Config.sol:39]: Actually no particular reason for that, it was just because I had created the interface of Config and used it in ScheduledPaymentModule before I created the implementation of Config, do you think it's better to use the autogenerated getter?

File: contracts/ScheduledPaymentModule.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (FadhlanR) [reply contracts/ScheduledPaymentModule.sol]: updated

File: contracts/ScheduledPaymentModule.sol [hunk 4]
```diff
@@ -281,6 +279,29 @@ contract ScheduledPaymentModule is Module {
...
+        uint256 _lastPaidAt
+    ) public view returns (bool) {
+        (uint256 year, uint256 month,) = BokkyPooBahsDateTimeLibrary
+            ._daysToDate(
+                block.timestamp / BokkyPooBahsDateTimeLibrary.SECONDS_PER_DAY
+            );
+        uint256 recursDate = BokkyPooBahsDateTimeLibrary.timestampFromDate(
+            year,
+            month,
+            recursDayOfMonth
```
[the remark this answers is not in the document]
Comment (FadhlanR) [reply contracts/ScheduledPaymentModule.sol]: updated

File: contracts/ScheduledPaymentModule.sol [hunk 5]
```diff
@@ -154,8 +154,10 @@ contract ScheduledPaymentModule is Module {
         if (!spHashes.contains(spHash)) revert UnknownHash(spHash);
         // 1 minute is buffer to protect against miners gaming block time
         // The recommended time for POW consensus finality is 1 minute
-        if (block.timestamp < payAt.add(1 minutes))
-            revert InvalidPeriod(spHash);
+        if (
+            block.timestamp < payAt.add(1 minutes) ||
+            block.timestamp > payAt.add(IConfig(config).getValidForDays())
```
[the remark this answers is not in the document]
Comment (FadhlanR) [reply contracts/ScheduledPaymentModule.sol]: updated to days instead of seconds

File: contracts/ScheduledPaymentModule.sol [hunk 6]
```diff
@@ -281,6 +279,29 @@ contract ScheduledPaymentModule is Module {
...
+            year,
+            month,
+            recursDayOfMonth
+        );
+
+        return
+            block.timestamp.sub(_lastPaidAt) >= 28 days && //recursDayOfMonth value range 1-28
+            block.timestamp >= recursDate &&
+            block.timestamp <=
+            recursDate.add(IConfig(config).getValidForDays()) &&
```
[the remark this answers is not in the document]
Comment (FadhlanR) [reply contracts/ScheduledPaymentModule.sol]: updated to days instead of seconds

File: test/Config.spec.ts (hunk 1 above)
Comment (FadhlanR) [reply test/Config.spec.ts]: updated

File: contracts/Config.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (alex-cardstack) [reply contracts/Config.sol]: No strong preference, this is clear and simple, maybe it uses a small amount less bytecode for the contract but I'm not sure of that

**Review by alex-cardstack — APPROVED** [review]

**Review by jurgenwerk — APPROVED** [review]

File: contracts/Config.sol [hunk 7]
```diff
@@ -5,23 +5,25 @@ import "./interfaces/IConfig.sol";
 import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
 
 contract Config is IConfig, OwnableUpgradeable {
-
     event ConfigSetup(address crankAddress, address feeReceiver);
 
     address private crankAddress;
     address private feeReceiver;
+    uint8 private validForDays;
```
Comment (habdelra) [inline contracts/Config.sol:12]: it feels odd that the state that we hold is the less fine grained time unit. We could still keep the outside API such that when you setup the contract you specify the "validFor" in units of days, but internally we could store as seconds. Be using the more fine grained unit within the contract that prevents the need from having to convert to from days to seconds when performing crank calls--which lowers over all gas usage. There is still the need to convert from days to seconds in the setup, but that would be a one time call.

File: contracts/Config.sol [hunk 8]
```diff
@@ -9,7 +9,7 @@ contract Config is IConfig, OwnableUpgradeable {
 
     address private crankAddress;
     address private feeReceiver;
-    uint8 private validForDays;
+    uint256 private validForSeconds;
```
Comment (habdelra) [inline contracts/Config.sol:12]: why bother making this private if you have an external function for returning this very value? it just adds extra bloat to the contract

File: contracts/Config.sol [hunk 9]
```diff
@@ -9,7 +9,7 @@ contract Config is IConfig, OwnableUpgradeable {
 
     address private crankAddress;
     address private feeReceiver;
```
Comment (habdelra) [inline contracts/Config.sol:11]: this should be public--no need for the function for returning this value

File: contracts/Config.sol [hunk 10]
```diff
@@ -9,7 +9,7 @@ contract Config is IConfig, OwnableUpgradeable {
 
     address private crankAddress;
```
[the remark this answers is not in the document]
Comment (habdelra) [reply contracts/Config.sol:10]: Generally we follow this kind of pattern in javascript (private property public getter) for a class property that is readonly to the outside world since in javascript encapsulation is much more difficult to enforce at runtime. However this kind of exploit is impossible in solidity since you can only change contract properties within the contract scope, not from outside of it.

https://stackoverflow.com/questions/49449474/is-it-possible-to-modify-a-variable-value-from-another-contract

File: contracts/Config.sol (hunk 7 above)
Comment (FadhlanR) [reply contracts/Config.sol]: updated

File: contracts/Config.sol (hunk 8 above)
Comment (FadhlanR) [reply contracts/Config.sol]: updated

File: contracts/Config.sol (hunk 9 above)
Comment (FadhlanR) [reply contracts/Config.sol]: updated

File: contracts/Config.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (FadhlanR) [reply contracts/Config.sol]: updated

**Review by habdelra — APPROVED** [review]
